linera-execution/src/execution.rs

```diff
@@ -231,27 +271,79 @@ where
                 tracing::error!("Ignoring error reported by user application: {message}");
                 RawExecutionResult::default()
             }
-            (runtime_result, call_result) => {
+            (runtime_result, execution_result) => {
                 runtime_result?;
-                call_result?
+                execution_result?
             }
         };
+
         // Set the authenticated signer to be used in outgoing messages.
-        result.authenticated_signer = signer;
-        let runtime_counts = runtime.runtime_counts();
+        execution_result.authenticated_signer = signer;
+        let runtime_counts = runtime.runtime_counts().await;
         let balance = self.system.balance.get_mut();
         tracker.update_limits(balance, policy, runtime_counts)?;
 
         // Check that applications were correctly stacked and unstacked.
         assert_eq!(applications.len(), 1);
         assert_eq!(applications[0].id, application_id);
-        // Make sure to declare the application first for all recipients of the user
-        // execution result.
+
+        // Update externally-visible results.
+        results.push(ExecutionResult::User(application_id, execution_result));
+        // Check that all sessions were properly closed.
+        ensure!(
+            session_manager.states.is_empty(),
+            ExecutionError::SessionWasNotClosed
+        );
+        Ok(results)
+    }
+
+    async fn run_user_action_with_synchronous_runtime(
+        &mut self,
+        application_id: UserApplicationId,
+        chain_id: ChainId,
+        action: UserAction,
+        policy: &ResourceControlPolicy,
+        tracker: &mut ResourceTracker,
+    ) -> Result<Vec<ExecutionResult>, ExecutionError> {
+        let balance = self.system.balance.get();
+        let runtime_limits = tracker.limits(policy, balance);
+        let initial_remaining_fuel = policy.remaining_fuel(*balance);
+        let (execution_state_sender, mut execution_state_receiver) =
+            futures::channel::mpsc::unbounded();
+        let execution_results_future = tokio::task::spawn_blocking(move || {
+            ContractSyncRuntime::run_action(
+                execution_state_sender,
+                application_id,
+                chain_id,
+                runtime_limits,
+                initial_remaining_fuel,
+                action,
+            )
+        });
+        while let Some(request) = execution_state_receiver.next().await {
+            self.handle_request(request).await?;
+        }
+        let (execution_results, runtime_counts) = execution_results_future.await??;
```

Comment:
Rust n00b question: what do the double question mark `??` do here? I thought it was for when we had two `Result` types within another, but that doesn't seem to be the case here 🤔

Reply:
There are two results to unpack. One from the closure and one from spawn_blocking

Reply:
Ah I see. Thank you!